### PR TITLE
Add HttpServerTracing adapter for Brave tracer

### PR DESCRIPTION
**Pull request by dmexe**
Changes:

* Add `HttpServerTracing` which decorates existing HTTP handlers
* Add `routeName` property to `HttpServerResponse`

At this moment client tracing cannot be implemented due to #558 

Related to #539

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@dmexe Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-netty&pullRequestId=622)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-netty&pullRequestId=622) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@dmexe Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-netty&pullRequestId=622)!

**Comment by smaldini** [comment]
That's awesome, we'll look at it. Maybe some input from @adriancole as well ?

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ *         .build();
+ *   HttpServerTracing serverTracing = HttpServerTracing.create(tracing);
+ *
+ *   DisposableServer c = HttpServer.create()
+ *     .port(0)
+ *     .handle(serverTracing.andThen(myHandler))
+ *     .bindNow();
+ * </pre></code>
+ */
+public class HttpServerTracing {
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java]: Usually, what we do is something like this...

```java
ReactorNettyTracing reactorTracing = ReactorNettyTracing.create(httpTracing);

....
 DisposableServer c = HttpServer.create()
    .port(0)
    .handle(reactorTracing.serverHandler().andThen(myHandler))
    .bindNow();

```

It isn't required, but more conventional as the top-level type (in this case `ReactorNettyTracing`) is supplied by the object graph. If this library is is server-only, it is probably also fine what you've done.

File: src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import com.sun.istack.internal.NotNull;
+import java.util.function.Consumer;
+import reactor.util.annotation.NonNull;
+import reactor.util.annotation.Nullable;
+import reactor.util.context.Context;
+
+/**
+ * Span related methods for {@link Context}.
+ */
+public class TracingContext {
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java]: nit if this is not a user-type, hide it?

File: src/test/java/reactor/netty/instrumentation/zipkin/HttpServerTracingTest.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+import reactor.netty.http.server.HttpServer;
+import reactor.netty.http.server.HttpServerRequest;
+import reactor.netty.http.server.HttpServerResponse;
+import reactor.netty.http.server.HttpServerRoutes;
+import reactor.test.StepVerifier;
+import zipkin2.Annotation;
+import zipkin2.Span;
+import zipkin2.Span.Kind;
+
+public class HttpServerTracingTest {
```
Comment (adriancole) [inline src/test/java/reactor/netty/instrumentation/zipkin/HttpServerTracingTest.java]: it is easy to miss edge cases, eventhough ad-hoc tests are good for understanding things locally. It would be nice to add the tests we use for all frameworks, to help understand what works vs doesn't.

ex https://github.com/openzipkin/brave/tree/master/instrumentation/vertx-web/src/test/java/brave/vertx/web

which implies a test dependency on
```
    <dependency>
      <groupId>io.zipkin.brave</groupId>
      <artifactId>brave-instrumentation-http-tests</artifactId>
      <scope>test</scope>
</dependency>
```

Comment (dmexe) [reply src/test/java/reactor/netty/instrumentation/zipkin/HttpServerTracingTest.java]: Thanks, `http-tests` found some mistakes

File: src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java (hunk 2 above)
Comment (dmexe) [reply src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java]: It's a decorator for reactive `Context` which provides span related methods for user level code.

for example:

```
Mono.subscriberContext()
  .flatMap(ctx -> {
    TracingContext.of(ctx).span(span -> span.annotate("log message"));
    return ...;
  });

**Comment by smaldini** [comment]
That PR is pretty cool, it made me think about either providing a better extension option on client/server level as we are improving builder API for 0.9 like : 

```java
      server.extension(TracingExtension)
                .handle(normalHandler)
       //...
```

Or in that case directly offering a `trace` method with a default support for Brave. That would require us to extract the necessary hook used here. We do the same with metrics() in reactor-core `Flux#metrics and micrometers.

I'm also thinking about routeName and how this will be symmetric for a similar support on the client side. Should we use a generic concept of tag instead that could be reused multiple times and by various frameworks (e.g. micrometer too) ? We could tag any operation.

File: src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java (hunk 2 above)
Comment (adriancole) [reply src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java]: cool. heh so, I'll nag you then :) javadoc the example? :D

**Comment by adriancole** [comment]
@smaldini on tag.. depends on the goal.

route is a mildly generic term, but not as generic as operation name (aka span name)
https://github.com/openzipkin/brave/blob/master/instrumentation/http/src/main/java/brave/http/HttpAdapter.java#L50

We use it where possible to create a route-based span name. This makes things "just work" with no custom logic.
https://github.com/openzipkin/brave/blob/master/instrumentation/http/src/main/java/brave/http/HttpParser.java#L63

OTOH an aribtrary tag could also be a custom name parsing input, and/or also a normal tag

I really like this PR also, btw!

**Comment by shakuzen** [comment]
> Should we use a generic concept of tag instead that could be reused multiple times and by various frameworks (e.g. micrometer too) ?

It definitely helps for correlation if metrics and tracing use the same value (to a lesser degree, name also) for like concepts like HTTP route. This was one of the reasons [Haystack](https://expediadotcom.github.io/haystack/) gave for generating metrics after-the-fact from tracing data - it ensures the tags match.

File: src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java (hunk 2 above)
Comment (dmexe) [reply src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java]: I've added Javadoc examples for each user level methods

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java (hunk 1 above)
Comment (dmexe) [reply src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java]: I think it isn't necessary, `brave.http.HttpTracing` is a good abstraction for both client and server configurations and customizations, so `HttpServerTracing` does nothing except wrap an existing hander to provide tracing functionality.

**Comment by dmexe** [comment]
> it made me think about either providing a better extension option
It would be great

> Or in that case directly offering a trace method with a default support for Brave. That would require us to extract the necessary hook used here. We do the same with reactor-core Flux#metrics and micrometer.

Direct depending on brave isn't good, opposed to micrometer which is a facade library and provides an API interface. Also, micrometer allows end-user using various metrics libraries (or just disable it). 

> I'm also thinking about routeName and how this will be symmetric for a similar support on the client side.

routeName for clients and servers should be an internal property. In the future, it would be easily constructed for a server from `HttpServerRoutes` predicates and for a `HttpClient` by extending an `uri` method to accept URI templates, for example `httpClient.uri("/items/{itemId}", itemId).get()`

**Comment by violetagg** [comment]
> Changes:
> 
> * Add `HttpServerTracing` which decorates existing HTTP handlers
> * Add `routeName` property to `HttpServerResponse`
> 
> At this moment client tracing cannot be implemented due to #558
> 
> Related to #539

@dmexe #558 is fixed and available with `0.8.7.BUILD-SNAPSHOT`

**Comment by adriancole** [comment]
hi, one site is asking about this now..

File: build.gradle [hunk 4]
```diff
@@ -62,6 +62,7 @@ ext {
   // Libraries
   nettyVersion = '4.1.34.Final'
   jacksonDatabindVersion = '2.5.1'
+  braveVersion = "5.6.1"
```
Comment (adriancole) [inline build.gradle]: switch to 5.8.0 for some more powerful things.. especially allows secondary sampling

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpCarrier.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+package reactor.netty.instrumentation.zipkin;
+
+import brave.propagation.Propagation.Getter;
+import brave.propagation.Propagation.Setter;
+import io.netty.handler.codec.http.HttpHeaders;
+
+/**
+ * HTTP headers injector/extractor
+ */
+class HttpCarrier implements Getter<HttpHeaders, String>, Setter<HttpHeaders, String> {
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpCarrier.java]: In 5.8 this type is no longer needed

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpClientAdapter.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package reactor.netty.instrumentation.zipkin;
+
+import reactor.netty.http.client.HttpClientRequest;
+import reactor.netty.http.client.HttpClientResponse;
+
+/**
+ * Brave HTTP adapter for client requests
+ */
+class HttpClientAdapter extends brave.http.HttpClientAdapter<HttpClientRequest, HttpClientResponse> {
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpClientAdapter.java]: In 5.8, implement brave's HttpClientRequest and HttpClientResponse instead.

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpServerAdapter.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+import brave.Span;
+import java.net.InetSocketAddress;
+import reactor.netty.http.server.HttpServerRequest;
+import reactor.netty.http.server.HttpServerResponse;
+
+/**
+ * Brave HTTP adapter for server requests
+ */
+class HttpServerAdapter extends brave.http.HttpServerAdapter<HttpServerRequest, HttpServerResponse> {
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpServerAdapter.java]: In 5.8, implement brave's HttpServerRequest and HttpServerResponse instead.

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  private final HttpServerHandler<HttpServerRequest, HttpServerResponse> handler;
+
+  /**
+   * Create a new tracing adapter using preconfigured {@link HttpTracing} instance.
+   */
+  private HttpServerTracing(@NonNull HttpTracing httpTracing) {
+    Objects.requireNonNull(httpTracing, "httpTracing cannot be null");
+
+    this.extractor = httpTracing.tracing().propagation().extractor(GETTER);
+    this.handler = HttpServerHandler.create(httpTracing, new HttpServerAdapter());
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java]: in 5.8, no extractor, just `HttpServerHandler.create(httpTracing) `

File: src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  > andThen(
+      @NonNull BiFunction<
+          ? super HttpServerRequest,
+          ? super HttpServerResponse,
+          ? extends Publisher<Void>
+      > next) {
+    Objects.requireNonNull(next, "next cannot be null");
+
+    return (request, response) -> {
+      final Span span = handler.handleReceive(extractor, request.requestHeaders(), request);
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/HttpServerTracing.java]: in 5.8, no extractor, just `handler.handleReceive(new YourHttpRequest(request));`

File: src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.netty.instrumentation.zipkin;
+
+import brave.Span;
```
Comment (adriancole) [inline src/main/java/reactor/netty/instrumentation/zipkin/TracingContext.java]: If this is a user type, please don't expose Span, expose SpanCustomizer instead. We have code notes why.

**Comment by violetagg** [comment]
This PR is superseded by #1307 
@dmexe @adriancole @smaldini PTAL

**Comment by violetagg** [comment]
Closing this as superseded by #1307
